oak_restricted_kernel/src/virtio_console.rs

```diff
@@ -0,0 +1,203 @@
+                Layout::from_size_align(pages * 0x1000, 0x1000).unwrap(),
+            );
+        }
+
+        0
+    }
+
+    fn phys_to_virt(paddr: virtio_drivers::PhysAddr) -> virtio_drivers::VirtAddr {
+        ADDRESS_TRANSLATOR
+            .get()
+            .unwrap()
+            .translate_physical(PhysAddr::new(paddr as u64))
+            .unwrap()
+            .as_u64() as usize
+    }
+
+    fn virt_to_phys(vaddr: virtio_drivers::VirtAddr) -> virtio_drivers::PhysAddr {
+        ADDRESS_TRANSLATOR
+            .get()
+            .unwrap()
+            .translate_virtual(VirtAddr::new(vaddr as u64))
+            .unwrap()
+            .as_u64() as usize
+    }
+}
+
+#[repr(transparent)]
+pub struct MmioConsoleChannel<'a> {
+    inner: Spinlock<VirtIOConsole<'a, OakHal, MmioTransport>>,
+}
+
+// Safety: for now, this is safe as we don't have threads in our kernel.
+unsafe impl Sync for MmioConsoleChannel<'_> {}
+unsafe impl Send for MmioConsoleChannel<'_> {}
+
+impl oak_virtio::Read for MmioConsoleChannel<'_> {
+    fn read(&mut self, data: &mut [u8]) -> anyhow::Result<()> {
+        let mut console = self.inner.lock();
+
+        let len = data.len();
+        let mut count = 0;
+        while count < len {
+            if let Some(char) = console
+                .recv(true)
+                .map_err(|err| anyhow!("Virtio console read error: {:?}", err))?
+            {
+                data[count] = char;
+                count = count + 1;
+            } else {
+                // If we didn't have any data to read, try acking any pending interrupts. We don't
+                // have general support for interrupts yet, so we need to do it blindly here.
+                console.ack_interrupt().unwrap();
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl oak_virtio::Write for MmioConsoleChannel<'_> {
+    fn write(&mut self, data: &[u8]) -> anyhow::Result<()> {
+        let mut console = self.inner.lock();
+
+        for char in data {
+            console
+                .send(*char)
+                .map_err(|err| anyhow!("Virtio console write error: {:?}", err))?;
+        }
+
+        Ok(())
+    }
+
+    fn flush(&mut self) -> anyhow::Result<()> {
+        Ok(())
+    }
+}
+
+fn find_memory_range(device: &AcpiDevice, ctx: &mut AmlContext) -> Option<(PhysAddr, PhysAddr)> {
+    for resource in device.crs(ctx).unwrap().unwrap() {
+        match resource {
+            Resource::MemoryRange(MemoryRangeDescriptor::FixedLocation {
+                is_writable: _,
+                base_address,
+                range_length,
+            }) => {
+                return Some((
+                    PhysAddr::new(base_address as u64),
+                    PhysAddr::new((base_address + range_length) as u64),
+                ))
+            }
+            _ => continue,
+        }
+    }
+    None
+}
+
+pub fn get_console_channel<'a, X: Translator>(
+    translator: &X,
```

Comment:
Maybe `translator: &impl Translator` works here?

Reply:
Indeed it does, thanks.